Add support for KIP-554 (Describe/Alter SCRAM Credentials)

michael-redpanda opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

This PR adds support for [KIP-554](https://cwiki.apache.org/confluence/display/KAFKA/KIP-554%3A+Add+Broker-side+SCRAM+Config+API) which adds the following Kafka APIs:

* AlterUserScramCredentials
  * Add SCRAM credentials
  * Alter existing SCRAM credentials
  * Delete existing SCRAM credentials
* DescribeUserScramCredentials
  * Describes either specific credentials or all credentials

todo:

- [ ] Add client tests

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

### Features

* Redpanda now supports [KIP-554](https://cwiki.apache.org/confluence/display/KAFKA/KIP-554%3A+Add+Broker-side+SCRAM+Config+API)

michael-redpanda committed:
k/p/schemata: Added describe user scram credentials

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
kafka/tests: Moved functions to common fixture

Will be used in future testing of other kafka messages

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
k/protocol: Added error codes for SCRAM messages

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
kafka: Added security utils for SCRAM messages

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
kafka: Introduce describe_user_scram_credentials

Initial commit that defines handlers for the
DescribeUserScramCredentials Kafka API.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
kafka: Added describe all user implementation

Added implementation to describe_user_scram_credentials to list all
known SCRAM users.

Also added unit tests to validate behavior both using and without using
authorization.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
kafka: Finished implementation of describe_user_scram_credentials

Added implementation to return specified users.  Will return errors when
provided either a missing user, an empty user, or if a user is provided
more than once.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
k/p/schemata: Added alter user scram credentials

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
kafka: Introduce alter_user_scram_credentials

Defines handler for the AlterUserScramCredentials API handler

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
security: Added make_credentials for pre-salted password

AlterUserScramCredentials supplies pre-salted password when creating or
modifying a user.  Added a new make_credentials method for SCRAM
algorithms to use this.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
kafka: Implemented alter user scram credentials

Added implementation for alter user scram credentials.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
redpanda/test: Ability to select SCRAM type

Signed-off-by: Michael Boquard <michael@redpanda.com>

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--e9e11fae-cc40-4cda-a5aa-a3167db4633d-->
<!--TEST RESULTS-->
<!--3c36a8fb-e07b-409f-97dc-c7b21e270866-->
<!--TEST RESULTS-->
<!--f5d697ad-3a95-4389-b57f-2db51fc8e710-->
<!--TEST RESULTS-->
<!--933f66c7-302e-4cea-ac8f-08095e016a7a-->
<!--TEST RESULTS-->
<!--c9d65901-e0b9-4133-be7f-945bb4420ddd-->
<!--TEST RESULTS-->
<!--4c6e2e83-6204-4728-8645-d1243b2bbb35-->
<!--TEST RESULTS-->
<!--e26a559c-e18c-4c4e-b021-aa9b29b9da6b-->
<!--TEST RESULTS-->
<!--49f779bc-4f31-448e-a09a-6b1fea669621-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61180">build#61180</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:----------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| gtest_raft_rpunit.gtest_raft_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/61180#019499f4-6fe0-429a-8cdb-db114381d8da | FLAKY | 1/2 |
| kafka_server_rpfixture.kafka_server_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/61180#019499f4-6fe0-40eb-b6e0-886dfbf5abc9 | FAIL | 0/2 |
| kafka_server_rpfixture.kafka_server_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/61180#019499f4-6fe0-429a-8cdb-db114381d8da | FAIL | 0/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64943">build#64943</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| kafka_server_rpfixture.kafka_server_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/64943#019662ce-b548-40be-a6e5-1f69b382f70f | FAIL | 0/2 |
| kafka_server_rpfixture.kafka_server_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/64943#019662ce-b549-46ab-8630-9df42a57ffab | FAIL | 0/2 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.with_chunked_compaction=False.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/64943#01966319-cb3f-491f-9cc8-622c329d58c7 | FLAKY | 20/21 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65016">build#65016</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| kafka_server_rpfixture.kafka_server_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/65016#0196683e-52fe-4df4-b1ca-9f5b2b9cbfe2 | FAIL | 0/2 |
| kafka_server_rpfixture.kafka_server_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/65016#0196683e-5300-427f-8297-09cbb5b84d34 | FAIL | 0/2 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=False.mixed_versions=False.with_tiered_storage=False.with_iceberg=True.with_chunked_compaction=True.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/65016#01966882-659b-4f25-861d-6506a84dc0bf | FLAKY | 20/21 |
| rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/65016#01966882-659b-4f25-861d-6506a84dc0bf | FLAKY | 17/21 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65108">build#65108</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| kafka_server_rpfixture.kafka_server_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/65108#01966d02-f239-4e83-8be1-045866a7060e | FAIL | 0/2 |
| kafka_server_rpfixture.kafka_server_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/65108#01966d02-f23b-475f-ad67-ff33eb04c3ff | FAIL | 0/2 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_conflicting_names | ducktape | https://buildkite.com/redpanda/redpanda/builds/65108#01966d5b-d6f5-47a4-8cc0-2d697624397d | FLAKY | 6/21 |
| rptest.transactions.consumer_offsets_test.VerifyConsumerOffsetsThruUpgrades.test_consumer_group_offsets.versions_to_upgrade=2 | ducktape | https://buildkite.com/redpanda/redpanda/builds/65108#01966d48-106a-4547-98df-d518d92e1095 | FLAKY | 20/21 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65124">build#65124</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| kafka_server_rpfixture.kafka_server_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/65124#01966e5c-9d87-4f81-abf2-e552243ca0cd | FAIL | 0/2 |
| kafka_server_rpfixture.kafka_server_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/65124#01966e5c-9d88-4ac2-9f3d-288d3fbd98dc | FAIL | 0/2 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_conflicting_names | ducktape | https://buildkite.com/redpanda/redpanda/builds/65124#01966e9c-9a73-4551-97ab-5343aa0bb815 | FLAKY | 5/21 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_conflicting_names | ducktape | https://buildkite.com/redpanda/redpanda/builds/65124#01966ea2-84e9-4a3f-a0eb-0d8149d0a2b7 | FLAKY | 6/21 |
| rptest.tests.datalake.disk_budget_test.DatalakeDiskUsageTest.test_idle_finish.num_partitions=10.concurrent_translations=4.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/65124#01966ea2-84e8-40e2-b794-81763091472a | FAIL | 0/1 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.with_chunked_compaction=True.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/65124#01966ea2-84e8-40e2-b794-81763091472a | FLAKY | 20/21 |
| rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/65124#01966ea2-84e8-40e2-b794-81763091472a | FLAKY | 14/21 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65540">build#65540</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.compaction_test.CompactionTest.test_compaction.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/65540#0196a1ab-2753-4c13-9984-000b984780e3 | FLAKY | 20/21 |
| rptest.tests.node_restart_probe_test.NodePostRestartProbeTest.post_restart_probe_test | ducktape | https://buildkite.com/redpanda/redpanda/builds/65540#0196a1ab-2754-4565-a625-93587c370c0e | FLAKY | 20/21 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65593">build#65593</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.partition_balancer_test.PartitionBalancerTest.test_unavailable_nodes | ducktape | https://buildkite.com/redpanda/redpanda/builds/65593#0196a79b-2f28-4d7f-a336-0b878f03de99 | FLAKY | 17/21 |
| rptest.tests.quota_management_test.QuotaManagementTest.test_error_handling | ducktape | https://buildkite.com/redpanda/redpanda/builds/65593#0196a79b-2f2a-415b-a691-89c24aa8ca29 | FAIL | 0/21 |
| rptest.tests.quota_management_test.QuotaManagementTest.test_error_handling | ducktape | https://buildkite.com/redpanda/redpanda/builds/65593#0196a7b2-d3a0-44a9-8539-77cad86ec947 | FAIL | 0/21 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/65593#0196a79b-2f28-4d7f-a336-0b878f03de99 | FLAKY | 7/21 |
| rptest.tests.usage_test.UsageTestCloudStorageMetrics.test_usage_manager_cloud_storage | ducktape | https://buildkite.com/redpanda/redpanda/builds/65593#0196a7b2-d3a0-44a9-8539-77cad86ec947 | FLAKY | 20/21 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65629">build#65629</a></summary>


| test_id                                                                                                                                                                       | test_kind   | job_url                                                                                   | test_status   | passed   |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_restore_partition_spec.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65629#0196aaa7-6f4b-4204-9489-4f3d1687c8d7 | FLAKY         | 20/21    |
| rptest.tests.partition_balancer_test.PartitionBalancerTest.test_unavailable_nodes                                                                                             | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65629#0196aaa7-6f46-4b26-af11-3cbaebf72092 | FLAKY         | 12/21    |
| rptest.tests.prefix_truncate_recovery_test.PrefixTruncateRecoveryTest.test_prefix_truncate_recovery.acks=-1.start_empty=False                                                 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65629#0196aab1-f242-4db2-b4b6-c82e22c4a974 | FLAKY         | 20/21    |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic                                                                                               | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65629#0196aaa7-6f46-4b26-af11-3cbaebf72092 | FLAKY         | 11/21    |
| rptest.tests.scram_test.SaslPlainTest.test_plain_authn.client_type=ClientType.PYTHON_RDKAFKA.scram_type=SCRAM-SHA-512.sasl_plain_enabled=False                                | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65629#0196aaa7-6f47-4867-9913-7c540d1c0c36 | FLAKY         | 20/21    |
| rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False                                                                          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65629#0196aab1-f242-4db2-b4b6-c82e22c4a974 | FLAKY         | 11/21    |
</details>
<!--/body-->

dotnwat reviewed (commented):
epic

BenPope commented on src/v/kafka/protocol/errors.h:
```diff
@@ -229,6 +229,12 @@ enum class error_code : int16_t {
     throttling_quota_exceeded = 89,
+    // A request illegally referred to a resource that does not exist.
+    resource_not_found = 91,
+    // A request illegally referredf to the same resource twice.
```
```suggestion
    // A request illegally referred to the same resource twice.
```

BenPope commented on src/v/kafka/protocol/describe_user_scram_credentials.h:
```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
```
```suggestion
 * Copyright 2025 Redpanda Data, Inc.

michael-redpanda replied:
ha this probably because I started working on this last year

BenPope commented on src/v/kafka/server/handlers/describe_user_scram_credentials.cc:
```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
```
```suggestion
 * Copyright 2025 Redpanda Data, Inc.

BenPope commented on src/v/kafka/server/handlers/describe_user_scram_credentials.h:
```diff
@@ -0,0 +1,19 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
```
```suggestion
 * Copyright 2025 Redpanda Data, Inc.

BenPope commented on src/v/kafka/server/handlers/describe_user_scram_credentials.cc:
```diff
@@ -42,9 +67,30 @@ ss::future<response_ptr> describe_user_scram_credentials_handler::handle(
+          -> bool {
+            return ss::visit(
+              t.second, [](const security::scram_credential&) { return true; });
+        };
```
Do you really want ephemeral credentials and roles?

michael-redpanda replied:
Good call out, no I don't

BenPope commented on src/v/kafka/server/handlers/describe_user_scram_credentials.cc:
```diff
@@ -42,9 +67,30 @@ ss::future<response_ptr> describe_user_scram_credentials_handler::handle(
+        res.data.results.emplace_back(describe_user_scram_credentials_result{
+          .user = user,
+          .credential_infos = {scram_credential_to_credential_info(c)}});
+    };
```
The scope of this can be inside list_all_users, but it might be clearer to just inline it.

BenPope commented on src/v/kafka/server/handlers/describe_user_scram_credentials.cc:
```diff
@@ -37,6 +39,101 @@ scram_credential_to_credential_info(const security::scram_credential& c) {
+                    item.name));
+            }
+        } catch (...) {
+            // store.get<T> throws if the user exists but is not of type T
```
Maybe use `std::holds_alternative<security::scram_credential>` to unify the error path?

michael-redpanda replied:
The `std::get<T>` on the variant occurs within the `credential_store::get<T>` method, I don't have access to the underlying variant.  Alternatively I guess i could use the `credential_store::range`

pgellert replied:
How about introducing a helper in `credential_store`, something like this?
```
    template<typename T>
    bool contains_type(const credential_user& name) const {
        if (auto it = _credentials.find(name); it != _credentials.end()) {
            return std::holds_alternative<T>(name);
        }
        return false;
    }
```

michael-redpanda replied:
I'm going to just change `get()` to instead return std::nullopt if the type isn't found.  Right now there's only a single credential type.

BenPope commented on src/v/kafka/server/handlers/describe_user_scram_credentials.cc:
```diff
@@ -37,6 +39,101 @@ scram_credential_to_credential_info(const security::scram_credential& c) {
+              fmt::format(
+                "Cannot describe SCRAM credentials for the same user twice in "
+                "a "
+                "single request: ",
```
nit: formatting

BenPope commented on src/v/kafka/protocol/alter_user_scram_credentials.h:
```diff
@@ -0,0 +1,44 @@
+#include "kafka/protocol/schemata/alter_user_scram_credentials_response.h"
+#include "kafka/protocol/wire.h"
+namespace kafka {
+struct alter_user_scram_credentials_request final {
```
nitpick: A little vertical space never hurt anybody!

BenPope commented on src/v/kafka/server/handlers/alter_user_scram_credentials.cc:
```diff
@@ -0,0 +1,109 @@
+template<typename Iter, typename ErrIter>
+void populate_results_with_error(
+  Iter begin,
+  Iter end,
```
nitpick: You could use ranges here.

BenPope commented on src/v/kafka/server/handlers/alter_user_scram_credentials.cc:
```diff
@@ -0,0 +1,109 @@
+          request.data,
+          res.data,
+          error_code::policy_violation,
+          "Recovery mjode enabled");
```
typo

BenPope commented on src/v/kafka/server/handlers/describe_user_scram_credentials.cc:
```diff
@@ -0,0 +1,50 @@
+    if (unlikely(ctx.recovery_mode_enabled())) {
+        res.data.error_code = error_code::policy_violation;
+        res.data.error_message = "Recovery mode enabled";
+        co_return co_await ctx.respond(std::move(res));
```
This doesn't need to be a coroutine.

BenPope commented on src/v/kafka/server/handlers/alter_user_scram_credentials.cc:
```diff
@@ -0,0 +1,109 @@
+          error_code::policy_violation,
+          "Recovery mjode enabled");
+
+        co_return co_await ctx.respond(std::move(res));
```
This doesn't need to be a coroutine.

michael-redpanda replied:
Not at this commit no, but it does become one in future commits.  If this is confusing I can modify these commit to just return and then change them later.

BenPope commented on src/v/security/scram_algorithm.h:
```diff
@@ -260,6 +260,22 @@ class scram_algorithm {
+          std::move(storedkey),
+          iterations,
+          std::move(principal)};
+    }
```
These `make_credentials` functions could probably have a little refactoring applied to reduce duplication.

BenPope commented on src/v/redpanda/tests/fixture.h:
```diff
@@ -897,41 +897,44 @@ class redpanda_thread_fixture {
     void authn_kafka_client(
       kafka::client::transport& client,
       const ss::sstring& username,
       const ss::sstring& password) {
```
```suggestion
    template<typename Authenticator = security::scram_sha256_authenticator>
    void authn_kafka_client(
      kafka::client::transport& client,
      const ss::sstring& username,
      const ss::sstring& password) {
        using ScramMech = typename Authenticator::auth::scram;

michael-redpanda commented:
Force push [1bf45e4](https://github.com/redpanda-data/redpanda/commit/1bf45e485c79a864ddea13c14e684f2c3a147d71):

* Rebased off of dev

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/64943-->
<!--53d9741a-07fb-4e03-a1ce-2c7c2aad5b18-->
<!--title-->
## Retry command for [Build#64943](https://buildkite.com/redpanda/redpanda/builds/64943)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"enable_failures":true,"mixed_versions":false,"with_chunked_compaction":false,"with_iceberg":true,"with_tiered_storage":true}
```
<!--/body-->

michael-redpanda commented:
Force push [bec7be7](https://github.com/redpanda-data/redpanda/commit/bec7be78c6135572992c3e2329d6fc5c2bcaa3e2):

* Addressed PR comments

michael-redpanda commented:
Force push [1c0f5b1](https://github.com/redpanda-data/redpanda/commit/1c0f5b1ea9e8460937214c862942c9c6f3aae27c):

* Fixup to use ranges in `populate_results_with_error` in `alter_user_scram_credentials.cc`

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/65016-->
<!--97cc6f33-1eb9-484f-8694-d4279a529b77-->
<!--title-->
## Retry command for [Build#65016](https://buildkite.com/redpanda/redpanda/builds/65016)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"enable_failures":false,"mixed_versions":false,"with_chunked_compaction":true,"with_iceberg":true,"with_tiered_storage":false}
```
<!--/body-->

secpanda commented:
![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/pr-banner-default.svg)

### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devex-c3w/pr-checks/a31bbb52-6059-4b72-aae5-32a80eb0ec09?refs=e30b2c33db9e8dc23a8a20b077f18ce579076b32&source=prChecksComment)

:white_check_mark: **license/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devex-c3w/pr-checks/a31bbb52-6059-4b72-aae5-32a80eb0ec09/license?refs=e30b2c33db9e8dc23a8a20b077f18ce579076b32&source=prChecksComment)

secpanda commented:
![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/pr-banner-default.svg)

### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devprod/pr-checks/3e28a8ef-32af-4c7f-b920-1059f809abe2?refs=e30b2c33db9e8dc23a8a20b077f18ce579076b32&source=prChecksComment)

:white_check_mark: **license/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devprod/pr-checks/3e28a8ef-32af-4c7f-b920-1059f809abe2/license?refs=e30b2c33db9e8dc23a8a20b077f18ce579076b32&source=prChecksComment)

michael-redpanda commented:
Force push [b9f83ca](https://github.com/redpanda-data/redpanda/commit/b9f83ca78c6dae2ff2e4ad77d0089f03c5b5a2b4):

* Rebased to fix merge conflicts introduced by #25919

pgellert commented on src/v/kafka/server/handlers/details/security.cc:
```diff
@@ -0,0 +1,38 @@
+    case security::scram_algorithm_t::sha512:
+        return kafka::scram_mechanism::scram_sha_512;
+    }
+    __builtin_unreachable();
```
nit:
```suggestion
```

michael-redpanda replied:
This actually isn't even used so I can remove the whole thing

pgellert commented on src/v/kafka/server/handlers/details/security.cc:
```diff
@@ -0,0 +1,38 @@
+    case kafka::scram_mechanism::scram_sha_512:
+        return security::scram_algorithm_t::sha512;
+    }
+    return std::nullopt;
```
nit:
```suggestion
```

michael-redpanda replied:
I do want this return - this is converting the value sent on the wire to a `kafka::scram_mechanism`.  Though the spec calls out 0, 1, 2 for unknown, sha256, and sha512 respectively, it's possible a misbehaving client may send something else.

pgellert commented on src/v/security/scram_credential.h:
```diff
@@ -19,6 +19,11 @@
+    sha256 = 0,
+    sha512 = 1,
+};
+
```
nit: do the underlying type and the values matter here? I'm wondering if this should just be
```
enum class scram_algorithm_t {
    sha256,
    sha512,
};
```

pgellert commented on src/v/kafka/server/handlers/describe_user_scram_credentials.cc:
```diff
@@ -0,0 +1,50 @@
+
+    describe_user_scram_credentials_response res;
+
+    if (unlikely(ctx.recovery_mode_enabled())) {
+        res.data.error_code = error_code::policy_violation;
+        res.data.error_message = "Recovery mode enabled";
+        return ctx.respond(std::move(res));
+    }
```
Should this be disabled in recovery mode? Looking through the list of things that are supposed to work in recovery mode, I would expect all user-managing APIs to work: https://docs.redpanda.com/current/manage/recovery-mode/#enabled-functionality

michael-redpanda replied:
ah good point, it doesn't need to be

pgellert commented on src/v/kafka/server/tests/describe_user_scram_credentials_test.cc:
```diff
@@ -0,0 +1,326 @@
+      "exists_256",
+      security::scram_sha256::make_credentials(
+        "password_256", security::scram_sha256::min_iterations));
+    create_user(
+      "exists_512",
+      security::scram_sha256::make_credentials(
+        "password_512", security::scram_sha256::min_iterations));
```
nit:
```suggestion
    create_user(
      "exists_512",
      security::scram_sha256::make_credentials(
        "password_512", security::scram_sha512::min_iterations));
```

michael-redpanda replied:
No it's 256 (I'm mkaing a 256 credential here) - poor name choice on my part

pgellert replied:
Yeah, maybe worth renaming to `exists_1` and `exists_2` or similar if you end up pushing more changes. I see that it's correct as is, but this inconsistency between the name and the algorithm used makes it harder to follow the test.

pgellert commented on src/v/kafka/server/handlers/alter_user_scram_credentials.cc:
```diff
@@ -0,0 +1,98 @@
+
+    alter_user_scram_credentials_response res;
+
+    if (unlikely(ctx.recovery_mode_enabled())) {
```
Same comment about recovery mode applies.

pgellert commented on src/v/kafka/server/handlers/alter_user_scram_credentials.cc:
```diff
@@ -0,0 +1,400 @@
+          request.data.deletions,
+          std::back_inserter(res.data.results),
+          error_code::broker_not_available,
+          "Broker not available - audit system failure");
```
I think this could just be:
```suggestion
        populate_results_with_error(
          request.data,
          res.data,
          error_code::broker_not_available,
          "Broker not available - audit system failure");
```

pgellert commented on src/v/kafka/server/handlers/alter_user_scram_credentials.cc:
```diff
@@ -93,6 +253,148 @@ ss::future<response_ptr> alter_user_scram_credentials_handler::handle(
+              model::timeout_clock::now() + 5s);
+
+        } else {
+            vlog(klog.debug, "Created SCRAM credentials for user {}", u.name);
```
nit: `Created -> Creating`

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/65108-->
<!--3ffacccc-7fed-49a4-94ec-5087e79a4810-->
<!--cae940fc-59f3-4842-a43f-e85dbe3738b8-->
<!--title-->
## Retry command for [Build#65108](https://buildkite.com/redpanda/redpanda/builds/65108)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```

/ci-repeat 1
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_conflicting_names
tests/rptest/transactions/consumer_offsets_test.py::VerifyConsumerOffsetsThruUpgrades.test_consumer_group_offsets@{"versions_to_upgrade":2}
```
<!--/body-->

michael-redpanda committed:
security: Do not throw from credential_store::get

Return std::nullopt rather than throwing an exception of the credentials
being fetched are of a different type.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda commented:
Force push [40323fe](https://github.com/redpanda-data/redpanda/commit/40323fe2db6488c27def855b14c9d777354d1831) & [0645440](https://github.com/redpanda-data/redpanda/commit/0645440c7bce952f0b14fc074bbb5f5e6dcf8b2e):

* Addressed PR comments

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/65124-->
<!--ba227816-284a-451c-9ef6-11a2c1434040-->
<!--71c43d2a-df86-465a-b10d-228e6567a845-->
<!--fd2bdb6e-8160-461c-aad9-6dfff5945354-->
<!--title-->
## Retry command for [Build#65124](https://buildkite.com/redpanda/redpanda/builds/65124)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```


/ci-repeat 1
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_conflicting_names
tests/rptest/tests/datalake/disk_budget_test.py::DatalakeDiskUsageTest.test_idle_finish@{"cloud_storage_type":1,"concurrent_translations":4,"num_partitions":10}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"enable_failures":true,"mixed_versions":false,"with_chunked_compaction":true,"with_iceberg":true,"with_tiered_storage":true}
tests/rptest/tests/upgrade_test.py::UpgradeBackToBackTest.test_upgrade_with_all_workloads@{"single_upgrade":false}
```
<!--/body-->

pgellert commented on src/v/kafka/server/handlers/alter_user_scram_credentials.cc:
```diff
@@ -79,6 +239,148 @@ ss::future<response_ptr> alter_user_scram_credentials_handler::handle(
+      request.data.upsertions.end(),
+      std::back_inserter(res.data.results));
+
+    auto invalid_scram_name_it = validate_range_valid_scram_user_name(
+      upsert_begin,
+      upsert_valid_range_end,
+      std::back_inserter(res.data.results));
+    upsert_valid_range_end = invalid_scram_name_it;
```
We check if the user name is valid using `validate_range_valid_scram_user_name` only for upserts, but not for deletes. That seems reasonable, but I just wanted to confirm that this is intentional.

michael-redpanda replied:
yes - invalid scram usernames shouldn't exist so the caller should get "resource_not_found", but in case they do permit their deletion.

pgellert commented on src/v/security/credential_store.h:
```diff
@@ -53,7 +53,10 @@ class credential_store {
     auto get(const credential_user& name) const -> const std::optional<T> {
         if (auto it = _credentials.find(name); it != _credentials.end()) {
-            return std::get<T>(it->second);
+            return ss::visit(
+              it->second,
+              [](const T& cred) { return cred; },
+              [](const auto&) { return std::nullopt; });
```
Are there any code paths relying on the old behaviour?

michael-redpanda replied:
Doesn't appear so:

* https://github.com/redpanda-data/redpanda/blob/04b20e3c96c13ef4e2bdef3767bfa0a3f6851d83/src/v/redpanda/admin/security.cc#L105-L115
* https://github.com/redpanda-data/redpanda/blob/04b20e3c96c13ef4e2bdef3767bfa0a3f6851d83/src/v/security/plain_authenticator.cc#L84-L88
* https://github.com/redpanda-data/redpanda/blob/04b20e3c96c13ef4e2bdef3767bfa0a3f6851d83/src/v/security/request_auth.cc#L115-L125
* https://github.com/redpanda-data/redpanda/blob/04b20e3c96c13ef4e2bdef3767bfa0a3f6851d83/src/v/security/scram_authenticator.cc#L31-L35

As we currently only have one credential type, we never have had to handle the exceptional case.  If we do add one, I think this improves things as callers don't appear to handle the exception.

pgellert commented on src/v/kafka/server/handlers/alter_user_scram_credentials.cc:
```diff
@@ -0,0 +1,386 @@
+  Iter begin, Iter end, ErrIter out_it) {
+    using type = Iter::value_type;
+    auto valid_range_end = std::partition(begin, end, [](const type& item) {
+        return security::validate_scram_username(item.name());
```
Do we need to validate the username for control characters like on the admin API? https://github.com/redpanda-data/redpanda/blob/9e6328a7e147a78c2a9a99654edf8efbad2c50eb/src/v/redpanda/admin/security.cc#L463

michael-redpanda replied:
No the deserialization code path handles that:

https://github.com/redpanda-data/redpanda/blob/04b20e3c96c13ef4e2bdef3767bfa0a3f6851d83/src/v/kafka/protocol/wire.h#L83-L86

pgellert reviewed (commented):
What you have so far lgtm. I think if you add a couple of tests with `kafka-configs.sh` to verify the behaviour end-to-end, then this seems good to merge to me.

BenPope commented on src/v/kafka/protocol/errors.h:
```diff
@@ -229,6 +229,12 @@ enum class error_code : int16_t {
+    resource_not_found = 91,
+    // A request illegally referred to the same resource twice.
+    duplicate_resource = 92,
+    /// Requested credential would not meet criteria for acceptability
```
nitpick:
```suggestion
    // Requested credential would not meet criteria for acceptability

BenPope commented on src/v/kafka/protocol/types.cc:
```diff
@@ -127,4 +127,16 @@ std::ostream& operator<<(std::ostream& os, config_resource_operation t) {
+std::ostream& operator<<(std::ostream& os, scram_mechanism m) {
+    switch (m) {
+    case scram_mechanism::scram_sha_256:
+        return os << "SCRAM-SHA-256";
+    case scram_mechanism::scram_sha_512:
+        return os << "SCRAM-SHA-512";
```
nitpick: These could come from scram_shaxxx_authenticator::name (maybe that's a bit of a circular reference)

michael-redpanda replied:
yeah that would induce a circular reference to `//src/v/security`

dotnwat replied:
>yeah that would induce a circular reference to //src/v/security

sounds like the solution might be pluck out the scram bits into its own library within the security module. that will break the cycle

BenPope commented on src/v/kafka/server/BUILD:
```diff
@@ -212,6 +214,7 @@ redpanda_cc_library(
         "//src/v/hashing:xx",
         "//src/v/kafka/data:partition_proxy",
         "//src/v/kafka/protocol",
+        "//src/v/kafka/protocol:describe_user_scram_credentials",
```
nitpick: Added on wrong line (fixed in next commit)

BenPope commented on src/v/kafka/server/tests/describe_user_scram_credentials_test.cc:
```diff
@@ -0,0 +1,178 @@
+
+        BOOST_REQUIRE(saw_sha256 && saw_sha512);
+    }
+};
```
I think this would be more clearly expressed as a container of the two users, sorted, and then the containers compared.

BenPope commented on src/v/security/credential_store.h:
```diff
@@ -53,7 +53,10 @@ class credential_store {
     auto get(const credential_user& name) const -> const std::optional<T> {
         if (auto it = _credentials.find(name); it != _credentials.end()) {
-            return std::get<T>(it->second);
+            return ss::visit(
```
nitpick: In the commit message /s/through/throw/

BenPope commented on src/v/kafka/server/handlers/describe_user_scram_credentials.cc:
```diff
@@ -37,6 +39,90 @@ scram_credential_to_credential_info(const security::scram_credential& c) {
+    using type = typename Iter::value_type;
+    chunked_hash_map<std::string_view, uint32_t> freq;
+    freq.reserve(std::distance(begin, end));
+    for (const auto& r : boost::make_iterator_range(begin, end)) {
```
nitpick:
```suggestion
    for (const auto& r : std::ranges::subrange(begin, end)) {

BenPope commented on src/v/kafka/server/handlers/describe_user_scram_credentials.cc:
```diff
@@ -37,6 +39,90 @@ scram_credential_to_credential_info(const security::scram_credential& c) {
+    for (auto it = valid_range_end; it != end; it++) {
+        if (freq[it->name()] > 1) {
+            *out_it = generate_error(
+              *it,
```
I think the loop looks nicer like this, and then`name_count = 1`, rather doing the lookup again.
```suggestion
    for (auto& user : std::ranges::subrange(valid_range_end, end)) {
        auto& name_count = freq[user.name()];
        if (name_count > 1) {
```

BenPope commented on src/v/kafka/server/handlers/describe_user_scram_credentials.cc:
```diff
@@ -37,6 +39,90 @@ scram_credential_to_credential_info(const security::scram_credential& c) {
+              fmt::format(
+                "Cannot describe SCRAM credentials for the same user twice in "
+                "a single request: ",
+                it->name));
```
This is unused, there's no `{}` in the format. (also,` user.name()`)

michael-redpanda replied:
thanks, good catch

BenPope commented on src/v/kafka/server/handlers/alter_user_scram_credentials.cc:
```diff
@@ -0,0 +1,84 @@
+
+template<typename T>
+kafka::alter_user_scram_credentials_result
+generate_error(const T& item, kafka::error_code code, const ss::sstring& msg) {
```
I'd be temped to either take `ss::sstring` by value and move, or by `string_view` to avoid the temporary and copy.

The latter is used for describe.

BenPope commented on src/v/security/scram_algorithm.h:
```diff
@@ -260,6 +260,25 @@ class scram_algorithm {
           std::move(principal));
     }
 
+    /// Test method used to generate credentials and it returns the salted
+    /// password
```
Should this go with the tests?

michael-redpanda replied:
Do you mean this change should end up in a commit with the tests?

BenPope replied:
> Do you mean this change should end up in a commit with the tests?

Maybe somewhere under the test directory, but it's a minor point.

BenPope commented on src/v/kafka/server/handlers/alter_user_scram_credentials.cc:
```diff
@@ -45,10 +54,161 @@ void populate_results_with_error(
+    return valid_range_end;
+}
+
+template<typename Iter, typename ErrIter>
+Iter validate_range_valid_scram_user_name(
+  Iter begin, Iter end, ErrIter out_it) {
```
I wonder if it's worth implementing both of the scram handlers in the same compilation unit to reduce duplication of these helpers.

michael-redpanda replied:
Yes, but can we save that for a follow on PR?  Reworking the commits woudl be a bit of a PITA at the moment

BenPope commented on src/v/kafka/server/handlers/alter_user_scram_credentials.cc:
```diff
@@ -45,10 +54,161 @@ void populate_results_with_error(
+    auto valid_delete_range_end = std::partition(
+      delete_begin, delete_end, [&freq](const delete_type& item) {
+          return freq[item.name()] == 1;
+      });
```
nitpick (I'm sure you can think of a better name than `partition`):
```suggestion
    const auto partition = [&freq]<typename Iter>(Iter begin, Iter end) {
        return std::partition(begin, end, [&freq](const auto& item) {
            return freq[item.name()] == 1;
        });
    };
    auto valid_upsert_range_end = partition(upsert_begin, upsert_end);
    auto valid_delete_range_end = partition(delete_begin, delete_end);

BenPope commented on src/v/kafka/server/handlers/alter_user_scram_credentials.cc:
```diff
@@ -45,10 +54,161 @@ void populate_results_with_error(
+          return freq[item.name()] == 1;
+      });
+
+    // We only want to respone _once_ for every duplicated user, so we will
+    // iterate through the partitioned list and generate an error if we
+    // encounter a user with a frequency greater than 1.  Once added, reset the
+    // frequency to 1 to skip the user if found again
+    const auto insert_error = [&freq]<typename Iter, typename OutIter>(
+                                Iter begin, Iter end, OutIter out_it) {
+        for (auto it = begin; it != end; it++) {
+            if (freq[it->name()] > 1) {
+                *out_it = generate_error(
+                  *it,
+                  kafka::error_code::duplicate_resource,
+                  "A user credential cannot be altered twice in the same "
+                  "request");
+                freq[it->name()] = 1;
+                ++out_it;
+            }
+        }
+    };
```
Very similar to what's in describe.

michael-redpanda replied:
yeah I'll file a ticket to refactor this a bit after merge, sounds good?

BenPope commented on src/v/kafka/server/handlers/alter_user_scram_credentials.cc:
```diff
@@ -79,6 +239,148 @@ ss::future<response_ptr> alter_user_scram_credentials_handler::handle(
+                  return creds.stored_key().size()
+                         == security::scram_sha512::key_size;
+              }
+          } catch (...) {
+              return false;
+          }
```
Is this still needed? Can it be more specific?

BenPope commented on src/v/kafka/server/handlers/alter_user_scram_credentials.cc:
```diff
@@ -79,6 +239,148 @@ ss::future<response_ptr> alter_user_scram_credentials_handler::handle(
+              }
+              const auto& creds = maybe_creds.value();
+              switch (mech) {
+              case security::scram_algorithm_t::sha256:
+                  return creds.stored_key().size()
+                         == security::scram_sha256::key_size;
+              case security::scram_algorithm_t::sha512:
+                  return creds.stored_key().size()
+                         == security::scram_sha512::key_size;
```
I thought I saw a utility function for this somewhere?

BenPope commented on src/v/kafka/server/handlers/alter_user_scram_credentials.cc:
```diff
@@ -45,10 +54,161 @@ void populate_results_with_error(
+  DeleteIter delete_begin,
+  DeleteIter delete_end,
+  ErrIter out_it) {
+    using upsert_type = typename UpsertIter::value_type;
+    using delete_type = typename DeleteIter::value_type;
```
nitpick: I don't think these are used.

michael-redpanda replied:
they are, but after refactoring, no longer

michael-redpanda commented:
[Force push](https://github.com/redpanda-data/redpanda/compare/0645440c7bce952f0b14fc074bbb5f5e6dcf8b2e..b11fb1a1148914e7f638bdb9729ac7cd56be218a):

* Rebased off dev

michael-redpanda commented:
Force push:

* Update to address PR comments
* FIxed wrong name being used in `generator.py`

michael-redpanda commented:
Force push:

* Rebased off dev

michael-redpanda commented:
> What you have so far lgtm. I think if you add a couple of tests with `kafka-configs.sh` to verify the behaviour end-to-end, then this seems good to merge to me.

This actually may be an issue - since we don't currently implement principal based throttling, `kafka-configs` errors out

michael-redpanda commented:
Force push:

* Addressed comments from AI

michael-redpanda commented on src/v/kafka/server/handlers/alter_user_scram_credentials.cc:
```diff
@@ -0,0 +1,374 @@
+    upsert_valid_range_end = upsert_dup_it;
+    deletions_valid_range_end = delete_dup_it;
+
+    const auto make_creds = [](const scram_credential_upsertion& u) {
```
There's no issue here - we check before this point that we have a valid mechanism in the upsert and we error compilation if there are missing case statements in a switch.

michael-redpanda replied:
There's no issue here - we check before this point that we have a valid mechanism in the upsert and we error compilation if there are missing case statements in a switch.

michael-redpanda committed:
dt: Add describe_user to kcli tools

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
kafka: Work-around describing user client quotas

This is a work around for the fact that we don't support user based
quotas yet.  The `kafka-configs` tool will error out when attempting to
describe a principal when the describe quotas handler returns
"unsupported_version" when it encounters a request to describe a user
quota.  This is a problem with implementing support for KIP-554, as
`kafka-configs` is a common kafka tool used to manager users.

As the alter path will still error out if asked to add quotas for a
user, this is an acceptable work around to not return anything when
describing quotas for a user.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
dt: Added tests to validate KIP-554 using kafka-configs

Signed-off-by: Michael Boquard <michael@redpanda.com>

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/65593-->
<!--f858f031-8b4d-4274-8101-36f601e47876-->
<!--a1f5abb3-673b-4966-8dfc-7be3058c8f42-->
<!--title-->
## Retry command for [Build#65593](https://buildkite.com/redpanda/redpanda/builds/65593)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```

/ci-repeat 1
tests/rptest/tests/quota_management_test.py::QuotaManagementTest.test_error_handling
```
<!--/body-->

michael-redpanda commented:
Force push:

* Fixed quota error test

pgellert reviewed (approved):
lgtm

BenPope reviewed (approved):
LGTM

dotnwat commented on src/v/kafka/server/handlers/describe_user_scram_credentials.cc:
```diff
@@ -0,0 +1,172 @@
+      .user = item.name, .error_code = code, .error_message = ss::sstring{msg}};
+}
+
+template<typename Iter, typename ErrIter>
+Iter validate_range_no_empty_users(Iter begin, Iter end, ErrIter out_it) {
```
is there something that is causing the need for so much templated code?
